### DISCUSSION TITLE
Custom controller implementation

### DISCUSSION THREAD

**Discussion by LukeWeger** [bitcraze/discussions#455] [Q&A]
Hello,

I am working on implementing a custom controller to the crazyflie. I have read up on how to do this and at this point have copied the existing PID controller and saved it as a new file, done the same with the PID header file, added all of the lines in the controller.c file to include this new controller, and updated the kbuild with the new controller file. I wanted to see if this new controller could be recognized by the crazyflie so I changed the default controller to my custom one within the controller.c file, and within the cfclient console it initializes with this new controller but instantly changes back to the PID controller. The same thing happens if I change the default controller to Mellinger as well where it will initialize with that and change to the PID controller. Am I missing the logic somewhere that is changing the controller to the PID controller regardless of what the default is set to?

Thanks for any assistance

**Comment by krichardsson** [comment]
If you want to force your controller to be used at start up, you should set it [here](https://github.com/bitcraze/crazyflie-firmware/blob/479464f2073a76b81830acf83b734847a7582b1a/src/modules/src/controller.c#L57)

The "normal" way of changing controller is otherwise to use the [`estimator.controllerType` parameter](https://github.com/bitcraze/crazyflie-firmware/blob/479464f2073a76b81830acf83b734847a7582b1a/src/modules/src/stabilizer.c#L368) which will [trigger the change in the stabilizer loop](https://github.com/bitcraze/crazyflie-firmware/blob/479464f2073a76b81830acf83b734847a7582b1a/src/modules/src/stabilizer.c#L278-L282)
